src/pykeen/training/schlichtkrull_sampler.py

```diff
@@ -132,3 +140,58 @@ def __iter__(self):  # noqa: D105
 
     def __len__(self):  # noqa: D105
         return self.num_batches_per_epoch
+
+
+class SubGraphInstances(Instances[BatchType], ABC):
+    """A dataset of subgraph samples."""
+
+    def __init__(
+        self,
+        *,
+        mapped_triples: MappedTriples,
+        sub_graph_size: int,
+        **kwargs,
+    ) -> None:
+        """
+        Initialize the subgraph dataset.
+
+        :param mapped_triples:
+            the ID-based triples
+        :param sub_graph_size:
+            the size (=number of triples) of the individual subgraph samples
+        :param kwargs:
+            additional keyword based arguments passed to super.__init__
+        """
+        super().__init__(**kwargs)
+        self.graph_sampler = GraphSampler(
+            mapped_triples=mapped_triples,
+            num_samples=sub_graph_size,
+        )
+
+
+class SLCWASubGraphInstances(SLCWAInstances, SubGraphInstances[SLCWABatchType]):
+    """SLCWA subgraph instances."""
+
+    def __init__(
+        self,
+        *,
+        mapped_triples: MappedTriples,
+        sub_graph_size: int,
+    ):
+        """
+        Initialize the subgraph dataset for SLCWA instances.
+
+        :param mapped_triples:
+            the ID-based triples.
+        :param sub_graph_size:
+            the size (=number of triples) of the individual subgraph samples
+        """
+        SLCWAInstances.__init__(self, mapped_triples=mapped_triples)
+        SubGraphInstances.__init__(self, mapped_triples=mapped_triples, sub_graph_size=sub_graph_size)
+
+    def __len__(self) -> int:  # noqa: D105
+        # is already batched!
+        return super().__len__() // self.graph_sampler.num_samples
+
+    def __getitem__(self, item: int) -> MappedTriples:  # noqa: D105
+        return torch.stack([SLCWAInstances.__getitem__(self, idx) for idx in list(self.graph_sampler)], dim=0)
```

Comment:
why do you have to call `list(self.graph_sampler)` instead of just iterating directly?

Reply:
there is no need :sweat_smile: 

83f9db1